### PR TITLE
[dune] [test-suite] switch to Dune based test-suite setup

### LINKED ISSUE #5990: `coqchk` test suite is not strong enough
See #1066 for more context, but basically we should check basic `coqchk` output status and printing functionality in the test-suite.

### LINKED ISSUE #6118: "make check" failures on coq-makefile test-suite
I observed that `make check` (from the main Makefile) fails on the `coq-makefile` test suite. It is not clear what happens but I suspect it is because of the cascade of calls of `make` from `make` (could it be?). Observationally, the differences in `log` files are of the form:
```
Makefile:680: warning: undefined variable `MAKECMDGOALS'
Makefile:587: warning: undefined variable `FOR_PACK'
Makefile:395: warning: undefined variable `TGTS'
Makefile:627: warning: undefined variable `src/test_plugin_MLLIB_DEPENDENCIES'
Makefile:401: warning: undefined variable `GAL'
```
followed by incomplete results of `make install` (as far as I could judge, though without full certainty).

### LINKED ISSUE #7146: The test-suite makefile does not spot missing dependencies
#### Version

Building from master.

#### Operating system

Debian unstable, using opam, switch `4.06.1`.

#### Description of the problem

When running the test-suite, it fails, probably because of missing dependencies. Which should be spotted by the test-suite makefile.

I run:

```
make clean
./configure
# using all the defaults
make build
make test-suite
```

End output of `make test-suite`:
```sh
TEST      coq-makefile/quick2vo
make[3] : on entre dans le répertoire « /home/leo/Programmation/OCaml/coq/test-suite »
BUILDING SUMMARY FILE
FAILURES
0m01.19s  | Compilers/Z/Bounds/Pipeline                   0m01.19s  | Compilers/Z/Bounds/Pipeline                    | 0m01.04s  || +0m00.14s | +14.42%     coq-makefile/extend-subdirs/run.sh...Error!
    coq-makefile/arg/run.sh...Error!
    coq-makefile/compat-subdirs/run.sh...Error!
    coq-makefile/coqdoc1/run.sh...Error!
    coq-makefile/coqdoc2/run.sh...Error!
    coq-makefile/emptyprefix/run.sh...Error!
    coq-makefile/findlib-package/run.sh...Error!
    coq-makefile/latex1/run.sh...Error!
    coq-makefile/merlin1/run.sh...Error!
    coq-makefile/mlpack1/run.sh...Error!
    coq-makefile/mlpack2/run.sh...Error!
    coq-makefile/multiroot/run.sh...Error!
    coq-makefile/only/run.sh...Error!
    coq-makefile/plugin1/run.sh...Error!
    coq-makefile/plugin2/run.sh...Error!
    coq-makefile/plugin3/run.sh...Error!
    coq-makefile/quick2vo/run.sh...Error!
    coq-makefile/timing/run.sh...Error!
    coq-makefile/uninstall1/run.sh...Error!
    coq-makefile/uninstall2/run.sh...Error!
    coq-makefile/validate1/run.sh...Error!
    coq-makefile/vio2vo/run.sh...Error!
make[3]: *** [Makefile:194: report] Error 1
make[3] : on quitte le répertoire « /home/leo/Programmation/OCaml/coq/test-suite »
make[2]: *** [Makefile:112: all] Error 2
make[2] : on quitte le répertoire « /home/leo/Programmation/OCaml/coq/test-suite »
make[1]: *** [Makefile.build:552: test-suite] Error 2
make[1] : on quitte le répertoire « /home/leo/Programmation/OCaml/coq »
make: *** [Makefile:190: submake] Error 2
```

I open this issue following an advice on #6809.

Thanks.

### LINKED ISSUE #13328: Comfortable development workflow?
### I am not doing work because the build system is lame.

I have been attempting to contribute to Coq since a while _(#11910, #11911)_, but every time I come back to it, there is some frustrating problem with the build system that makes me sad enough to abandon the work. For example, [the last time I discovered that **unit tests are not really performed**][1] and have not been for who knows how long a time. _(Resolved in #12289.)_ What I mean to express by this preamble is that, while such issues may be avoided so habitually by the frequent contributors as not to even be apparent to them, the experience of building and hacking is of any quality but comfortable. I understand that there is never enough hands to perfect everything, and the build system is of the least priority. But perhaps **there is not enough hands because it is so unnecessarily difficult to get going** in the first place!

[1]: https://gitter.im/coq/coq?at=5eb713ff9f0c955d7db58015

### The build system is lame.

One thing that makes me sad is that there are two build systems that depend on each other in a subtle way. These systems are `make` and `dune`.
* I can run an individual test only with the `make` system.
* I can run `utop` in a sub-directory only with the `dune` system.

But in order to use `dune`, I must remove `make`'s build products!

    % dune utop
    Error: Multiple rules generated for _build/default/META.coq:
    - file present in source tree
    - <internal location>
    Hint: rm -f META.coq

This means that I must either run all tests or rebuild a large part of the project every time I make a smallest modification. Either takes a long time — as long as hours. This gives a poor experience and makes me sad.

#### For a wider view:

* Why should there even be two build systems at once, for as long as at least the half a year I am around?
* Why should the two officially supported build systems squabble over some build products that are so insignificant as to even be ignored by the version control?

This does not make any sense.

### Possible solutions.

* In a more modern language ecosystem, there would be a flexible and reliable test runner. Example: Haskell's `tasty`. We cannot have this because porting the current test suite anywhere would require manual rewrites that would introduce errors. Perhaps some parts of the test suite, such as the innumerable `success` and `failure` bunches, could be automatically put under a modern test runner? Is there even a modern test runner in the Ocaml ecosystem?
* For some reason, `make` must keep build products all over the project, and `dune` must complain about that and crash. Can either of this be solved?
* Or maybe I am missing some blessed, simple and beautiful work flow that everyone is using? Let us put it to a more prominent place in the developers' documentation!

### I really want to help, only give me leadership.

If a system is to work, there must be a vision and an architecture. Any system left to float by itself inevitably becomes a mess, which is what we have here. It is not efficient to fix small things here and there. My proposition: `make` must be deprecated as soon as possible, and `dune` should be taught to run individual tests with colourful output. If there is a way I can help make it happen, I have hours upon hours of free work to offer.

### LINKED ISSUE #15186: "make -f Makefile.dune test-suite" always does a full build, never an incremental build
For example, just run the make twice in a row.  It repeats the same tests, e.g.

```
TEST      bugs/opened/bug_3370.v
TEST      bugs/opened/bug_3277.v
TEST      bugs/opened/bug_3304.v
TEST      bugs/opened/bug_3889.v
TEST      bugs/opened/bug_3938.v
TEST      bugs/opened/bug_3312.v
TEST      bugs/opened/bug_6393.v
```

Presumably these tests are passing since there's no FAILED for them

### PR DESCRIPTION
**Pull request by Alizter**
The test-suite now works by generating Dune rules and running them.

Simply run `make test` to watch it run.

- [ ] Documentation
  - [x] test-suite README - In progress, see `test-suite/README.new.md` (will be renamed)
  - [ ] Update dev docs
  - [ ] Update CI docs
- [x] CI Issues:
  - [x] test-suite jobs are rebuilding main job
    - This seems to be because we pass a different COQLIB to make the test-suite. Ideally, we should just keep the whole of _build/ for the main job and continue without configuring COQLIB. 
    - test-suite jobs now no longer rely on build and are directly started.
- Platform support
  - [x] Linux
  - [x] MacOS
    - Complexity tests, output modulo time and misc are disabled
  - Windows
    - For now the test-suite is diabled on Windows
- [x] Native support
  - [x] Doing any `./configure` whatsoever prevents coqlib from being found
        + Only seems to be on some tests, just needs fixing. 
  - If native is available build with native targets (or should we just use coqnative?) No for now.
- [x] Async failures: Running coqtop and async causes issues, reported in https://github.com/coq/coq/issues/17149.

Continuation of https://github.com/coq/coq/pull/13364

Issues:
- fixes https://github.com/coq/coq/issues/7146
- fixes https://github.com/coq/coq/issues/15186
- fixes https://github.com/coq/coq/issues/13328
- fixes https://github.com/coq/coq/issues/6118
- fixes https://github.com/coq/coq/issues/5990 (we run coqchk alot more now)

### CODE REVIEW

**Comment by SkySkimmer** [comment]
What's the replacement for 
- run 1 test (before: `make success/bla.v.log`)
- run 1 output test and update the reference file (before: `make output/bla.v.log; mv output/bla.out.real output/bla.out`)
- run 1 test without compiling its stdlib dependencies (before: `make success/bla.v.log`)

?

**Comment by SkySkimmer** [comment]
>Most tests will run without requiring the full stdlib. If you build dune build @bla you will see it only builds what it needs.

I want to ignore what it needs.

**Comment by ejgallego** [comment]
> I want to ignore what it needs.

Why you want to run `coqc` in a stale environment? This can be done, but it is a bad idea.

**Comment by ejgallego** [comment]
Note that now tests depend really on what they need, for most tests it is just `coqc` and `Prelude.vo`, unless they also spec `-noinit`.

**Comment by SkySkimmer** [comment]
suppose I compile Reals.vo
then I change something in a tactic implementation (eg tactics.ml)
if I think the change doesn't matter for Reals.vo I want to run tests using the tactic and requiring reals without recompiling reals

**Comment by ejgallego** [comment]
> suppose I compile Reals.vo then I change something in a tactic implementation (eg tactics.ml) if I think the change doesn't matter for Reals.vo I want to run tests using the tactic and requiring reals without recompiling reals

I guess if you want to be a build system you can just call `coqc` too? Maybe have a script that runs just the test action, or allow the rules to be unsound?

But experience in large codebases has taught us that the gain is not worth the increase on cognitive load; this kind of manual setup only works for really small cliques / use cases. The consensus is that giving away soundness to gain on speed leads in the end to a doomed developer experience.

**Comment by Alizter** [comment]
@SkySkimmer I have added a way to run .v file tests without rebuilding their deps. As before when running a single test you can do
```sh
dune build @DiscrR
```
however if you want to build this test without having the dependencies rebuilt you can do
```sh
dune build @DiscrR.replay
```
This will create the same action however with a renamed file so that the targets don't interfere with the other Dune rules.

**Comment by SkySkimmer** [comment]
The CI test log seems pretty verbose, do we need to print the full commands?

**Comment by Alizter** [comment]
@SkySkimmer That's for debugging purposes at the moment.

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coq/doc-maintainers Could you take a look at my new documentation for the test-suite. I will move it from being `README.new.md` to `README.md` soon, but I kept it separate for reviewing purposes.

File: test-suite/README.new.md
Comment (Alizter) [inline test-suite/README.new.md:1]: @coq/doc-maintainers

**Review by jfehrle — CHANGES_REQUESTED** [review]
The doc needs considerable improvement so it reads like a reference work and is thus useful.  Developers should be able to find what they need without reading the entire doc.  The doc needs better organization, e.g.
- provide an introductory subsection and introductory sentences where needed
- put the most important info first
- group related information together and hierarchically in a very logical order

Not to be harsh, but it's a bit of a jigsaw puzzle--the reader needs to examine lots of text to find anything and there are a bunch of missing pieces, like introductions.  If you put in some effort I will review multiple revisions as necessary.  (I haven't dug too deeply so far.)

File: test-suite/README.md
Comment (Alizter) [reply test-suite/README.md:103]: These are leftovers from Emilio that I will remove.

File: Makefile
Comment (SkySkimmer) [inline Makefile:142]: Please don't hide these commands

File: Makefile
Comment (SkySkimmer) [inline Makefile:165]: Having empty dune files in the test suite seems like it will produce confusion.
Why not merge dunestrap and duneteststrap?

Comment (Alizter) [reply Makefile:165]: I originally had both rules generated at the same time, but this will slow down the main build. I therefore opted to make them independent. We might be able to merge dunestrap and duneteststrap.

In this case, it might be possible to remove touching these files here, I will try it out.

File: test-suite/README.new.md
Comment (SkySkimmer) [inline test-suite/README.new.md:201]: This errors on the coqchk step for me
~~~
Fatal Error: User error: The file DiscrR__replay_0c4a9caccfc716c5b3fd8a3828cdafef.vo contains library
              DiscrR and not library
              DiscrR__replay_0c4a9caccfc716c5b3fd8a3828cdafef
~~~

Comment (SkySkimmer) [reply test-suite/README.new.md:201]: Also it seems the replay aliases don't do the output check for output tests.

File: test-suite/README.new.md
Comment (SkySkimmer) [inline test-suite/README.new.md:168]: When tests in separate directories have the same name they end up sharing the alias.
Is it possible to have the alias be `@bugs/bug_1446` or some such thing instead?

File: test-suite/README.new.md
Comment (Alizter) [reply test-suite/README.new.md:201]: I don't currently have support for replaying coqchk. The fact that coqchk has to have the same name for the vo makes doing this difficult.

File: test-suite/README.new.md
Comment (Alizter) [reply test-suite/README.new.md:168]: That alias will already work.

File: test-suite/README.new.md
Comment (SkySkimmer) [reply test-suite/README.new.md:201]: also also the replay targets aren't compatible with tests where the name of the file matters, eg univbinders:
~~~
File "./UnivBinders__replay_ff11647412b7cd297f02f3631133cf59.v", line 64, characters 0-49:
Error: Undeclared universe UnivBinders.mono.uu.
~~~
I don't remember if coqc allows -topfile, alternatively maybe the replay files could be in a subdirectory .replay/ instead of using some hash suffix.

**Comment by Alizter** [comment]
@SkySkimmer I have added aliases `@vo @chk` for building vo and coqchk separately. When you do `dune build @test` only the vo test is run. To do coqchk do `dune build @test.chk`.

**Comment by SkySkimmer** [comment]
>When you do `dune build @test` only the vo test is run

why this choice?

**Comment by Alizter** [comment]
@SkySkimmer I mean `dune build @mytest`.

**Comment by Alizter** [comment]
@SkySkimmer You can now write `dune build @bug_5030` for the entire test (vo and check). `@bug_5030.vo` for vo and `@bug_5030.chk` for coqchk.

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@SkySkimmer the two step rules will require more time than I can give this PR at the moment.

There is currently a failure in async due to #17149. You can also see the error reporting for the CI in that job.

Will you assign for tomorrow?

**Review by ejgallego — APPROVED** [review]
A very impressive amount of work; this seems indeed a good checkpoint to keep refining on.

I think rule generation can be improved still quite a bit, while rule_gen itself looks pretty neat, coq_rules is a bit messy, we could:
- define the notion of test and test group algebraically
- there should be a single function that generates the coq rules, with the proper options, instead of the current duplication
- this function is called by the upper level test-generation fucntion
- we need to use the infrastructure for rule generation in stdlib, in particular `Path.t` and other stuff is critical
- there could be a lot more sharing in terms of computing deps and traversing dirs, tho the global overhead due to this is likely low

We need to pay attention to the shape of the rules, as there are two shapes depending on whether we are in legacy plugin method or in findlib method; that should actually become a parameter as while for the stdlib we don't care, the test suite should test both.

**Comment by Alizter** [comment]
@jfehrle I didn't mean to ignore your comments here. I will come back to the documentation at a later date so that I can have some more feedback from other members of the team when they are using it in master.

File: test-suite/README.md
Comment (ejgallego) [reply test-suite/README.md:103]: Actually let's put that in some other PR, I think I have a more up-to-date version of this doc.

Comment (ejgallego) [reply test-suite/README.md:103]: The TODO for this "document env vars for the CI / test-suite setup"

**Comment by SkySkimmer** [comment]
I would like other devs to test this and confirm that they can adapt before merging. cc @coq/pushers 

I also still rather dislike having to rerun tests to get the error.

**Review by JasonGross — CHANGES_REQUESTED** [review]
> Also it seems the replay aliases don't do the output check for output tests.

This seems like a blocker for this PR, right?  If an output test fails, I want to be able to know what's going on from the CI log

**Comment by Alizter** [comment]
@JasonGross Output tests are diffed so there failure will be shown immediately. All tests are rerun in the CI to show the error messages. When you encounter a test that fails you can run `dune build @nameoftest.debug` to get an error message and a backtrace. See the README.new.md.

**Comment by SkySkimmer** [comment]
Because the rule fails, so the targets get cleaned up including the log file. This is why we need the 2 step rule

**Comment by JasonGross** [comment]
> Because the rule fails, so the targets get cleaned up including the log file. This is why we need the 2 step rule

Is there no way to tell dune to keep the intermediate files?

**Comment by ejgallego** [comment]
Doing the 2-step stuff is not so complex tho, the thing here is that this PR is very large so we'd rather try to work on merging this a bit more incrementally too.

**Comment by ejgallego** [comment]
@Alizter will be busy the next weeks with his studies, he asked me to take over this; I'm happy to do so.

First we need an assignee, @SkySkimmer would you like to shepherd this?

**Comment by jfehrle** [comment]
> I didn't mean to ignore your comments here. I will come back to the documentation at a later date so that I can have some more feedback from other members of the team when they are using it in master.

That's not a good approach.  IIUC what you've done is incomplete: it addresses certain cases but not others.  You're leaving it to every developer to discover through experimentation what actually works and what's broken.  The doc can and should identify what's complete and what's not complete today.  It won't take you long to add that info.

**Comment by SkySkimmer** [comment]
I can assign but I insist on getting more reasonable errors before merging?

**Comment by ejgallego** [comment]
> I can assign but I insist on getting more reasonable errors before merging. (also I have holidays next week)

Sounds good to me.

**Comment by ejgallego** [comment]
There is the ITP dealine also soon, so I'll rebase and implement the better error stuff when you are back.

**Comment by gares** [comment]
bump.

**Comment by Zimmi48** [comment]
> > I didn't mean to ignore your comments here. I will come back to the documentation at a later date so that I can have some more feedback from other members of the team when they are using it in master.
> 
> That's not a good approach. IIUC what you've done is incomplete: it addresses certain cases but not others. You're leaving it to every developer to discover through experimentation what actually works and what's broken. The doc can and should identify what's complete and what's not complete today. It won't take you long to add that info.

If the doc can be improved before merging that's good, but if it is not, then it shouldn't be a blocker for this particular PR. The reason to grant an exception is that this PR has already been waiting for too long and it deserves the project's interests that it is not merged. Once it is merged, it can be the job of every developer to improve the doc after experimenting, asking questions, and discovering what doesn't work as expected. (This is not a user-facing feature, so we can tolerate that the doc is not as polished as it should be in such a case.)

**Comment by SkySkimmer** [comment]
NB: The bad errors are however blockers as far as I'm concerned

**Comment by ejgallego** [comment]
Much to my regret I won't be able to work on this PR as I said I'd try to do. So if someone wants to take over, feel free.

**Pull request closed without merging**
